{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TMvis\n",
    "# Database for Transmembrane Protein Annotation and Visualization\n",
    "## Table of contents\n",
    "* [Step 1. Install dependencies](#first-bullet)\n",
    "* [Step 2. Provide path to TMvis dataset](#second-bullet)\n",
    "* [Step 3. Explore TMvis in 3D](#third-bullet)\n",
    "    * [TMbed](#tmbed)\n",
    "    * [ANVIL](#anvil)\n",
    "    * [PPM3](#ppm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Install dependencies <a class=\"anchor\" id=\"first-bullet\"></a>\n",
    "Press `Run` button on the top of the Notebook to install the neccessary dependencies. Make sure that the cell did not output any errors (warnings are okay) before you continue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "# 3D visualization of protein molecules\n",
    "!{sys.executable} -m pip install py3Dmol\n",
    "# iPython widgets e.g. sliders\n",
    "!{sys.executable} -m pip install ipywidgets\n",
    "# fasta files handling\n",
    "!{sys.executable} -m pip install Bio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Provide path to a folder with TMvis dataset<a class=\"anchor\" id=\"second-bullet\"></a>\n",
    "Run the following cell to connect the Notebook to TMvis dataset. The dataset folder should have the following structure: \n",
    "```\n",
    "db                           \n",
    "|                                    \n",
    "├── alpha                                 \n",
    "│   │                                 \n",
    "│   ├── pLDDT90F1\n",
    "|   |   |\n",
    "|   |   └──A0A087X1C5                                 \n",
    "|   │       ├── A0A087X1C5.fasta                                 \n",
    "|   │       ├── AF-A0A087X1C5-F1-model_v2.pdb                                 \n",
    "|   │       ├── AF-A0A087X1C5-F1-model_v2.cif                                 \n",
    "|   │       ├── AF-A0A087X1C5-F1-model_v2_ANVIL.pdb                                 \n",
    "|   │       └── AF-A0A087X1C5-F1-model_v2_ppm.PDB                                 \n",
    "|   |       └── ...                                    \n",
    "└── beta                                 \n",
    "    └── pLDDT90F1\n",
    "        └── P45880 \n",
    "        └── ...\n",
    "```\n",
    "**Please make sure to provide a correct path to the directory where TMvis dataset was downloaded.**        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "KK7X9T44pWb7"
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "# [TODO] Provide path to folder with transmembrane proteins called db (without including the db folder itself)\n",
    "home_path = \"./../results/current/\"\n",
    "file_path = os.path.join(home_path, 'db/')\n",
    "\n",
    "# all Uniprot IDs of TMvis\n",
    "alpha = sorted(os.listdir(os.path.join(file_path, 'alpha/pLDDT90F1')))\n",
    "beta = sorted(os.listdir(os.path.join(file_path, 'beta/pLDDT90F1')))\n",
    "print(\"# proteins in total: \", len(alpha + beta))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Explore TMvis in 3D<a class=\"anchor\" id=\"third-bullet\"></a>\n",
    "Run the following cells to visualize transmembrane proteins (TMPs) embedded in a membrane. To select a protein - start typing in a Uniprot ID to a dedicated field under the cell, then available Uniprot IDs from the TMvis dataset will appear."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "X-ofaXbhZRVv"
   },
   "source": [
    "## Transmembrane Protein Prediction with Embeddings<a class=\"anchor\" id=\"tmbed\"></a>\n",
    "## TMbed \n",
    "A novel method predicting alpha helical and beta barrel TMPs, their transmembrane segments and overall inside/outside topology relative to the membrane. Utilizing embeddings generated by language models from natural language processing and trained on protein sequences, it can predict whole proteomes within an hour while maintaining a prediction accuracy equal to or better than methods using evolutionary information (sequence profiles or alignments)[3]. Furthermore, in this section you can compare TMbed predictions to ANVIL[1] or PPM3[2] membrane assignment.          \n",
    "                         \n",
    "**TMbed color code**     \n",
    "<font color='green'>IN --> OUT orientation</font>       \n",
    "<font color='blue'>OUT --> IN orientation</font>     \n",
    "                \n",
    "**ANVIL and PPM color code**      \n",
    "<font color='red'>membrane</font>  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 373
    },
    "id": "VSUUayp3Yv_o",
    "outputId": "7eb92f47-ff71-4a70-85ad-5721f8debd66"
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import py3Dmol\n",
    "import glob\n",
    "import ipywidgets as widgets\n",
    "\n",
    "def show_tmbed(uniprotID, overlay_with=[\"ppm3\", \"anvil\"], membr_opacity=0.7):\n",
    "    if overlay_with == 'anvil':\n",
    "        if uniprotID in alpha:\n",
    "                pdb_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ANVIL.pdb'\n",
    "                fasta_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/{uniprotID}.fasta'\n",
    "        else:\n",
    "                pdb_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ANVIL.pdb'\n",
    "                fasta_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/{uniprotID}.fasta'\n",
    "    else:\n",
    "        # ppm3\n",
    "        if uniprotID in alpha:\n",
    "                pdb_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ppm.PDB'\n",
    "                fasta_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/{uniprotID}.fasta'\n",
    "        else:\n",
    "                pdb_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ppm.PDB'\n",
    "                fasta_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/{uniprotID}.fasta'\n",
    "\n",
    "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
    "\n",
    "    seq_obj = list(SeqIO.parse(fasta_file, \"fasta\"))[0]\n",
    "\n",
    "  # FASTA file contains ID, protein sequence and transmembrane residue prediction (3 lines)\n",
    "  # len(protein sequence) = len(transmembrane prediction) = num(residues)\n",
    "  # SeqIO parses everything from the second line after > as a sequence -> aa sequence and\n",
    "  # transmembrane prediction are captured together and needed to be splited from each other\n",
    "    seq_len = len(seq_obj.seq)\n",
    "    middle = seq_len // 2\n",
    "    tm_bed_seq = seq_obj.seq[middle:seq_len]\n",
    "\n",
    "    if overlay_with == 'anvil':\n",
    "        with open(pdb_file) as ifile:\n",
    "            system = \"\".join([x for x in ifile])\n",
    "            view.addModelsAsFrames(system)\n",
    "            num_atom = 0\n",
    "            for line in system.split(\"\\n\"):\n",
    "                split = line.split()\n",
    "                # skip comments in the PDB file\n",
    "                if (len(split) == 0) or (split[0] not in [\"ATOM\", \"HETATM90002\", \"HETATM90001\"]):\n",
    "                      continue\n",
    "                else: \n",
    "                    # membrane atoms definition\n",
    "                    if split[0] == \"HETATM90002\" or split[0] == \"HETATM90001\":\n",
    "                        view.setStyle({'model': -1, 'resn': 'DUM'}, {\"sphere\": {'color': 'red', 'opacity': membr_opacity}})\n",
    "                    else:\n",
    "                        # protein definition\n",
    "                        # extract the atoms residue\n",
    "                        res_num = split[5]\n",
    "                        res_type = tm_bed_seq [int(res_num) - 1]\n",
    "                        if res_type == '.':\n",
    "                              atom_color = 'grey'\n",
    "                        elif res_type in ['B', 'H']:\n",
    "                              atom_color = 'green'\n",
    "                        elif res_type in ['b', 'h']:\n",
    "                              atom_color = 'blue'\n",
    "                        else:\n",
    "                              atom_color = 'pink'\n",
    "                    num_atom += 1\n",
    "                    view.setStyle({'model': -1, 'serial': num_atom}, {\"cartoon\": {'color': atom_color}})\n",
    "    else:\n",
    "          with open(pdb_file) as ifile:\n",
    "            system = \"\".join([x for x in ifile])\n",
    "            view.addModelsAsFrames(system)\n",
    "            num_atom = 0\n",
    "            for line in system.split(\"\\n\"):\n",
    "                split = line.split()\n",
    "                # skip comments\n",
    "                if (len(split) == 0) or (split[0] not in [\"ATOM\", \"HETATM\"]):\n",
    "                    continue\n",
    "                else: \n",
    "                    if split[0] == \"HETATM\":\n",
    "                        view.setStyle({'model': -1, 'resn': 'DUM'}, {\"sphere\": {'color': 'red', 'opacity': membr_opacity}})\n",
    "                    else:\n",
    "                        # protein definition\n",
    "                        # extract the atoms residue\n",
    "                        res_num = split[5]\n",
    "                        res_type = tm_bed_seq [int(res_num) - 1]\n",
    "                        #print(res_type, \":\", res_num)\n",
    "                        if res_type == '.':\n",
    "                            atom_color = 'grey'\n",
    "                        elif res_type in ['B', 'H']:\n",
    "                            atom_color = 'green'\n",
    "                        elif res_type in ['b', 'h']:\n",
    "                            atom_color = 'blue'\n",
    "                        else:\n",
    "                            atom_color = 'pink'\n",
    "                        num_atom += 1\n",
    "                        view.setStyle({'model': -1, 'serial': num_atom}, {\"cartoon\": {'color': atom_color}})\n",
    "    view.zoomTo()\n",
    "    assert (len(tm_bed_seq) == int(res_num)), \"Number of residues is not equal to the length of the TMbed prediction string\"\n",
    "    return view.show()\n",
    "\n",
    "widgets.interact(show_tmbed, uniprotID=widgets.Combobox(\n",
    "    options=alpha+beta,\n",
    "    value=alpha[0]));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "gMwVFyoRCbK0"
   },
   "source": [
    "## Assignment aNd VIsualization of the Lipid bilayer<a class=\"anchor\" id=\"anvil\"></a>\n",
    "## ANVIL\n",
    "ANVIL[1] is aimed to assign membrane boundaries to a protein 3D structure, by using the spatial coordinates of the alpha carbons. It can also process coarse-grained models of protein structures. The algorithm follows an approach that treats the problem of membrane assignment as a binary classification. The software is implemented in Python and is freely available under a CeCILL License. In output, ANVIL writes a PDB file that contains the coordinates of the membrane (defined as heteroatoms and visualized here in red).           \n",
    "A protein is colored based on AlphaFold [4] pLDDT score or can be changed to rainbow color scheme. AlphaFold produces a per-residue confidence score (pLDDT) between 0 and 100. Some regions below 50 pLDDT may be unstructured in isolation.        \n",
    "\n",
    "<font color='blue'> Very high (pLDDT > 90)</font>      \n",
    "<font color='aqua'>  Confident (90 > pLDDT > 70)</font>     \n",
    "<font color='yellow'>  Low (70 > pLDDT > 50)</font>     \n",
    "<font color='orange'>  Very low (pLDDT < 50)</font>       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 374
    },
    "id": "MEugi-Bo_mh8",
    "outputId": "7db5bd22-2800-4c6e-a581-a4570757f490"
   },
   "outputs": [],
   "source": [
    "def show_anvil_pdb(uniprotID, show_sidechains=False, show_mainchains=False, show_heteroatoms=True, show_surface=True, color=\"lDDT\"):\n",
    "    \n",
    "    if uniprotID in alpha:\n",
    "        pdb_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ANVIL.pdb'\n",
    "    else:\n",
    "        pdb_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ANVIL.pdb'\n",
    "            \n",
    "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
    "    with open(pdb_file) as ifile:\n",
    "        system = \"\".join([x for x in ifile])\n",
    "        view.addModelsAsFrames(system)\n",
    "\n",
    "    if color == \"lDDT\":\n",
    "        view.setStyle({'model': -1}, {'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n",
    "        \n",
    "    elif color == \"rainbow\":\n",
    "        view.setStyle({'model': -1}, {'cartoon': {'color':'spectrum'}})\n",
    "        \n",
    "    if show_heteroatoms:\n",
    "        for line in system.split(\"\\n\"):\n",
    "            split = line.split()\n",
    "            if (len(split) == 0) or (split[0] not in [\"ATOM\", \"HETATM90002\", \"HETATM90001\"]):\n",
    "                  continue\n",
    "            if split[0] == \"HETATM90002\" or split[0] == \"HETATM90001\":\n",
    "                color = 'red'\n",
    "                view.setStyle({'model': -1, 'resn': 'DUM'}, {\"sphere\": {'color': color, 'opacity':0.7}})\n",
    "\n",
    "    if show_sidechains:\n",
    "        BB = ['C','O','N']\n",
    "        view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
    "                            {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "        view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]}, {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "        view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
    "                            {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    if show_mainchains:\n",
    "        BB = ['C','O','N','CA']\n",
    "        view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "\n",
    "    if show_surface:\n",
    "        view.addSurface(py3Dmol.VDW,{'opacity':0.8,'color':'grey'}, {'model': -1, 'resn': 'DUM', 'invert': 'True'})\n",
    "        \n",
    "    view.zoomTo()\n",
    "    return view.show()\n",
    "\n",
    "\n",
    "widgets.interact(show_anvil_pdb, uniprotID=widgets.Combobox(\n",
    "    options=alpha+beta,\n",
    "    value=alpha[0]));"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1KK0FImbEL4D"
   },
   "source": [
    "## Positioning of proteins in flat and curved membranes<a class=\"anchor\" id=\"ppm\"></a>\n",
    "## PPM3.0\n",
    "*   Positioning of transmembrane and peripheral proteins and peptides in membranes and micelles\n",
    "*   Calculation of binding modes and membrane affinity of proteins taking into account the influence of hydrophobic matching and curvature stress\n",
    "* prediction of protein-induced membrane deformations\n",
    "* calculations in flat artificial and natural membranes\n",
    "* calculations in curved membranes with adjustable radii of curvature\n",
    "* calculations in spherical micelles of different sizes\n",
    "* positioning of proteins spanning two parallel membranes (e.g. outer and inner membranes of Gram-negative bacteria)\n",
    "* approximation of deformed membrane surfaces by several (up to four) flat surfaces\n",
    "* using polarity profiles along the membrane normal for the DOPC (1,2-dioleoyl-sn-glycero-3-phosphocholine) bilayerusing hydrophobic thicknesses (± 5Å) of artificial lipid bilayers (experimental values) and of different biological membranes (average values obtained from the large-scale analysis of transmembrane proteins from the OPM database [2]       \n",
    "\n",
    "\n",
    "A protein is colored based on AlphaFold [4] pLDDT score or can be changed to rainbow color scheme. AlphaFold produces a per-residue confidence score (pLDDT) between 0 and 100. Some regions below 50 pLDDT may be unstructured in isolation.        \n",
    "\n",
    "<font color='blue'> Very high (pLDDT > 90)</font>      \n",
    "<font color='aqua'>  Confident (90 > pLDDT > 70)</font>     \n",
    "<font color='yellow'>  Low (70 > pLDDT > 50)</font>     \n",
    "<font color='orange'>  Very low (pLDDT < 50)</font>       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 497
    },
    "id": "HqybvQdwDSv3",
    "outputId": "f2408af8-35c8-4204-cfc8-4f40f362323c"
   },
   "outputs": [],
   "source": [
    "def show_ppm_pdb(uniprotID, show_sidechains=False, show_mainchains=False, show_heteroatoms=True, show_surface=True, color=\"lDDT\"):\n",
    "    if uniprotID in alpha:\n",
    "        pdb_file = f'{file_path}/alpha/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ppm.PDB'\n",
    "    else:\n",
    "        pdb_file = f'{file_path}/beta/pLDDT90F1/{uniprotID}/AF-{uniprotID}-F1-model_v2_ppm.PDB'\n",
    "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js')\n",
    "    \n",
    "    with open(pdb_file) as ifile:\n",
    "        system = \"\".join([x for x in ifile])\n",
    "    view.addModelsAsFrames(system)\n",
    "    \n",
    "    if color == \"lDDT\":\n",
    "        view.setStyle({'model': -1}, {'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n",
    "    elif color == \"rainbow\":\n",
    "        view.setStyle({'model': -1}, {'cartoon': {'color':'spectrum'}})\n",
    "    \n",
    "    if show_sidechains:\n",
    "        BB = ['C','O','N']\n",
    "        view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
    "                            {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "        view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]}, {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "        view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
    "                            {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})  \n",
    "    if show_mainchains:\n",
    "        BB = ['C','O','N','CA']\n",
    "        view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
    "    \n",
    "    if show_surface:\n",
    "        view.addSurface(py3Dmol.VDW,{'opacity':0.8,'color':'grey'}, {'model': -1, 'resn': 'DUM', 'invert': 'True'})\n",
    "    \n",
    "    if show_heteroatoms:\n",
    "        for line in system.split(\"\\n\"):\n",
    "            split = line.split()\n",
    "            if (len(split) == 0) or (split[0] not in [\"ATOM\", \"HETATM\"]):\n",
    "                  continue\n",
    "            if split[0] == \"HETATM\":\n",
    "                color = 'red'\n",
    "                view.setStyle({'model': -1, 'resn': 'DUM'}, {\"sphere\": {'color': color, 'opacity':0.7}})\n",
    "    view.animate({'loop': 'backAndForth'})\n",
    "    view.zoomTo()\n",
    "    return view.show()\n",
    "\n",
    "widgets.interact(show_ppm_pdb, uniprotID=widgets.Combobox(\n",
    "    options=alpha+beta,\n",
    "    value=alpha[0]));\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "1. Postic, Guillaume, Yassine Ghouzam, Vincent Guiraud, and Jean-Christophe Gelly. 2016. “Membrane Positioning for High- and Low-Resolution Protein Structures through a Binary Classification Approach.” Protein Engineering, Design & Selection: PEDS 29 (3): 87–91.        \n",
    "2. Lomize, Mikhail A., Irina D. Pogozheva, Hyeon Joo, Henry I. Mosberg, and Andrei L. Lomize. 2012. “OPM Database and PPM Web Server: Resources for Positioning of Proteins in Membranes.” Nucleic Acids Research 40 (Database issue): D370–76.                                 \n",
    "3. Bernhofer, Michael, and Burkhard Rost. 2022. “TMbed – Transmembrane Proteins Predicted through Language Model Embeddings.” bioRxiv. https://doi.org/10.1101/2022.06.12.495804.                            \n",
    "4. Tunyasuvunakool, Kathryn, Jonas Adler, Zachary Wu, Tim Green, Michal Zielinski, Augustin Žídek, Alex Bridgland, et al. 2021. “Highly Accurate Protein Structure Prediction for the Human Proteome.” Nature 596 (7873): 590–96.     \n",
    "5. Varadi, Mihaly, Stephen Anyango, Mandar Deshpande, Sreenath Nair, Cindy Natassia, Galabina Yordanova, David Yuan, et al. 2022. “AlphaFold Protein Structure Database: Massively Expanding the Structural Coverage of Protein-Sequence Space with High-Accuracy Models.” Nucleic Acids Research 50 (D1): D439–44."
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "TMvis.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
